test/test_torch.py

```diff
@@ -1293,5 +1293,28 @@ def test_synchronize_step_warning(self):
             assert 'optimizer.step(synchronize=True) called after optimizer.synchronize()' \
                 in str(ws[0].message)
 
+    def test_horovod_join(self):
+        """Test join."""
+        hvd.init()
+        rank = hvd.rank()
+        size = hvd.size()
+
+        dtypes = [torch.ByteTensor, torch.CharTensor, torch.ShortTensor,
+                  torch.IntTensor, torch.LongTensor, torch.FloatTensor, torch.DoubleTensor]
+        if _fp16_supported:
+            dtypes += [torch.HalfTensor]
+        if torch.cuda.is_available():
+            dtypes += [torch.cuda.ByteTensor, torch.cuda.CharTensor, torch.cuda.ShortTensor,
+                       torch.cuda.IntTensor, torch.cuda.LongTensor,
+                       torch.cuda.FloatTensor, torch.cuda.DoubleTensor]
+            if _fp16_supported:
+                dtypes += [torch.cuda.HalfTensor]
+        dims = [1, 2, 3]
+        for dtype, dim in itertools.product(dtypes, dims):
+            tensor = torch.FloatTensor(*([17] * dim)).fill_(1).mul_(rank)
+            tensor = self.cast_and_place(tensor, dtype)
+            ret = hvd.join()
```

Comment:
The join test should test the actual Join behavior - e.g. you can allreduce on rank 0, join on rank 1, and rank 1 will allreduce with zero dummy tensor.